File: ResultDotNet.CSharp.Specs/ReadmeSpec.cs
using System;
using System.Data;
using NUnit.Framework;
using PatternMatching;
using static LanguageExt.Prelude;
using ResultDotNet;
using static ResultDotNet.Result;

namespace ResultDotNet.CSharp.Specs {
  /// <summary>
  /// this just makes sure that everything in the README compiles
  /// </summary>
  public class ReadmeSpec {
    class Invoice { public double Total; public double NumberOfUnits; }

    Result<double, string> divide(double numerator, double denominator) =>
      (denominator == 0)
        ? Failure<double, string>("Cannot divide by 0!")
        : Success<double, string>(numerator / denominator);

    string pricePerUnitForDisplay(Invoice invoice) =>
      divide(invoice.Total, invoice.NumberOfUnits).Match(
        success: ppu => ppu.ToString(),
        failure: err => $"N/A: {err}");
    
    string pricePerUnitForDisplay2(Invoice invoice) {
      var ans = divide(invoice.Total, invoice.NumberOfUnits);
      if (ans is Result<double, string>.Success) 
        return (ans as Result<double, string>.Success).Item.ToString();
      else {
        var err = (ans as Result<double, string>.Failure).Item;
        return $"N/A: {err}";
      }
    }
    
    string pricePerUnitForDisplay3(Invoice invoice) =>
      Pattern.Match<Result<double, string>, string>(divide(invoice.Total, invoice.NumberOfUnits))
        .Case<Result<double, string>.Success>(div => div.Item.ToString())
        .Case<Result<double, string>.Failure>(err => $"N/A: {err.Item}");

    Result<double, string> pricePerUnit(Invoice invoice) => divide(invoice.Total, invoice.NumberOfUnits);
    Result<double, string> savingsPerUnit(Invoice invoice, double dollarsOff) =>
      pricePerUnit(invoice).Bind(ppu => divide(dollarsOff, ppu));

    Result<double, string> pricePerUnitWithDiscount(Invoice invoice, double dollarsOffPerUnit) =>
      pricePerUnit(invoice).Map(ppu => ppu - dollarsOffPerUnit);
    
    Result<double, string> savingsPerUnit2(Invoice invoice, double dollarsOff) =>
      Result.Bind(ppu => divide(dollarsOff, ppu),  pricePerUnit(invoice));

    Result<double, string> pricePerUnitWithDiscount2(Invoice invoice, double dollarsOffPerUnit) =>
      Result.Map(ppu => ppu - dollarsOffPerUnit, pricePerUnit(invoice));
    
    Invoice createInvoice(double total, double numberOfUnits) => 
      new Invoice() { Total = total, NumberOfUnits = numberOfUnits };

    Result<Invoice, string> createInvoice(Result<double, string> total, Result<double, string> numberOfUnits) =>
      Result.Map2(createInvoice, total, numberOfUnits);

    Result<DataTable, string> executeDatabaseQuery(string sql) => 
      Failure<DataTable, string>("no database");

    public void Test() {
      var sql = "";
      var logger = new { Log = act((string err) => {}) };
      Result<DataTable, string> result = executeDatabaseQuery(sql);
      result.IfFailure(err => logger.Log(err));

      result.Match(
        success: val => logger.Log($"DB query ran successfully: {sql}"),
        failure: err => logger.Log($"DB query FAILED: {sql}"));
    }
  }
}

File: ResultDotNet.Specs/ReadmeSpec.fs
namespace ResultDotNet.Specs

/// <summary>
/// this just makes sure that everything in the README compiles
/// </summary>
module ReadmeSpec =
  open System
  open System.Data
  open ResultDotNet
  open ResultDotNet.FSharp

  let divide (numerator:float) (denominator:float) =
    if denominator = 0.
    then Failure "Cannot divide by 0!"
    else Success (numerator / denominator)

  type Invoice = { Total:float; NumberOfUnits:float }

  let pricePerUnitForDisplay invoice =
    match divide invoice.Total invoice.NumberOfUnits with
    | Success ppu -> ppu.ToString()
    | Failure err -> "N/A: " + err

  let pricePerUnit invoice = divide invoice.Total invoice.NumberOfUnits
  let savingsPerUnit invoice dollarsOff =
    pricePerUnit invoice |> Result.bind (fun ppu -> divide dollarsOff ppu)
    // you could of course `pricePerUnit () |> Result.bind (divide dollarsOff)`
    // but I find it counterintuitive that dollarsOff would be the numerator with that syntax

  let pricePerUnitWithDiscount invoice dollarsOffPerUnit =
    pricePerUnit invoice |> Result.map (fun ppu -> ppu - dollarsOffPerUnit)

  let newInvoice total numberOfUnits = { Total = total; NumberOfUnits = numberOfUnits }

  let createInvoice2 (total:Result<double, string>) (numberOfUnits:Result<double, string>) =
    Result.map2 newInvoice total numberOfUnits

  let executeDatabaseQuery sql = 
    Result.Failure<DataTable, string> "no query"
    

  type Logger() = member this.Log err = ()
  let logger = new Logger ()
  let sql = ""

  let result:Result<DataTable, string> = executeDatabaseQuery sql
  result |> Result.ifFailure (logger.Log);

  let savingsPerUnit2 invoice dollarsOff =
    Result.expr {
      let! ppu = pricePerUnit invoice
      return! divide dollarsOff ppu
    }

  let pricePerUnitWithDiscount2 invoice dollarsOffPerUnit =
    Result.expr {
      let! ppu = pricePerUnit invoice
      return ppu - dollarsOffPerUnit
    }

  let createInvoice3 (total:Result<double, string>) (numberOfUnits:Result<double, string>) =
    Result.expr {
      let! t = total
      let! n = numberOfUnits
      return newInvoice t n
    }

